### ISSUE TITLE
Top Level LLK Api clean up

### ISSUE THREAD

**Issue by rtawfik01** [tenstorrent-metal/tt-metal#4732]
During WHB0/GS llk uplift, lots of issues were deferred to be able to commonalize llk_lib & common/inc directory. Here are the list of issues that should now be cleaned up.

### 1. Make per device template parameters/functions calls different in only one layer. ###

- [x] Issue: https://github.com/tenstorrent-metal/tt-metal/issues/2179
Note: ifdefs have been left for colmajor/rowmajor matmuls, and transpose/no transpose for matmuls & partial reload. That will be a separate issue.

**One method:** Clean up ifdefs in compute_kernel_api/,  and use same function headers in <device>/llk_api to make devices consistent. Example:

Remove the ifdefs here in tt_metal/include/compute_kernel_api/bcast.h:

```
    // unpacker
    #ifdef ARCH_GRAYSKULL
    UNPACK(( llk_unpack_A_init<BroadcastType::NONE, false, EltwiseBinaryReuseDestType::NONE>(0, 0, 255) ));
    #else
    UNPACK(( llk_unpack_A_init<BroadcastType::NONE, false, EltwiseBinaryReuseDestType::NONE>()  ));
    #endif
```

since the `llk_unpack_A_init` call is the same for both grayskull and wormhole_b0:
```
template <
    BroadcastType BType = BroadcastType::NONE,
    bool acc_to_dest = false,
    EltwiseBinaryReuseDestType binary_reuse_dest = EltwiseBinaryReuseDestType::NONE,
    bool unpack_to_dest = false /*not used*/>
inline void llk_unpack_A_init(
```

For wormhole_b0 unpack_to_dest is used, but it is not in grayskull, this might cause function parameter bloating.

**Second method:** Have separate compute_kernel_api files per device, clean the ifdefs & make all the function/template parameters only have parameters it uses:

```
template <
    BroadcastType BType = BroadcastType::NONE,
    bool acc_to_dest = false,
    EltwiseBinaryReuseDestType binary_reuse_dest = EltwiseBinaryReuseDestType::NONE>
inline void llk_unpack_A_init(
```

**Third method:** Have a generator mechanism to create template/function calls per device, but this might cause performance degradation.

### 2. Remove parameters that should not be exposed to upper layers ###

- [x] Issue: https://github.com/tenstorrent-metal/tt-metal/issues/2065

Some parameters are static in the stack. For example, DstSync::Half is whats used across the stack, so DstSync should not be exposed in the api layer

tt_metal/hw/ckernels/grayskull/metal/llk_api/llk_pack_api.h:
```

template <DstSync Dst, bool is_fp32_dest_acc_en = false>
inline void llk_pack_dest_section_done() {
    _llk_pack_dest_section_done_<Dst, is_fp32_dest_acc_en>();
}

template <DstSync Dst, DstTileFaceLayout FaceLayout, bool untilize = false>
inline void llk_init_packer_dest_offset_registers(const std::uint32_t pack_output = 16) {
    _llk_init_packer_dest_offset_registers_<Dst, FaceLayout, untilize>();
}
```

and can be hardcoded to `DstSync::Half`

### 3. Remove operand idx hardcoding ###

- [x] Issue: https://github.com/tenstorrent-metal/tt-metal/issues/1626

Multiple locations in the compute_kernel_api have operand indices hardcoded:

tt_metal/include/compute_kernel_api/bcast.h:
```
ALWI void sub_tiles_bcast_scalar_init_short()
{
    MATH(( llk_math_eltwise_binary_init<ELWSUB, BroadcastType::SCALAR, MATH_FIDELITY>() )); // TODO(AP)
    // FIXME: API Update needed in compute kernel?
    UNPACK(( llk_unpack_AB_init<BroadcastType::SCALAR>(0, 1) ));
}
```

That should be cleaned up where possible.

### 4. Inline vs Always_inline ###

Experiment with always/force inlines, see if it impacts performance, and if it affects code size. Then choose one method to be consistent across the stack. Currenrtly the api layer has a mix of both:

tt_metal/hw/ckernels/grayskull/metal/llk_api/llk_pack_api.h:
```
TT_ALWAYS_INLINE void llk_pack_relu_config(const std::uint32_t config) {
    _llk_pack_relu_config_(config);
}

inline void llk_pack_reconfig_l1_acc(const std::uint32_t enable) {
    _llk_pack_reconfig_l1_acc_(enable);
}
```

### 5. Add debug/watchers back to api layer ###

- [ ] During uplift, watchers had to be removed from llk_lib since it is a common dir across stacks, and watcher is only in metallium. However, debug watchers can be added back to the api layer. Example:

tt_metal/hw/ckernels/grayskull/common/inc/cmath_common.h: 
```
inline void wait_math_semaphores()
inline void wait_math_semaphores()
{
{
    // wait while math semaphore is on max, no room to write math results
    - DEBUG_STATUS('W', 'M', 'S', 'W');
    TTI_SEMWAIT(p_stall::STALL_MATH, semaphore::t6_sem(semaphore::MATH_PACK), p_stall::STALL_ON_MAX);
    TTI_SEMWAIT(p_stall::STALL_MATH, semaphore::t6_sem(semaphore::MATH_PACK), p_stall::STALL_ON_MAX);
   - DEBUG_STATUS('W', 'M', 'S', 'D');
```

The watchers got removed here: https://github.com/tenstorrent-metal/tt-metal/pull/4270/files

But they can added back in wrapper functions wherever required.

**Comment by rtawfik01** [comment]
@mywoodstock @yugaoTT @muthutt @DrJessop This is the master issue for all the clean up we discussed. You can decide how to separate the issues, and what kind of solution you think works better with overall dev in this stack

**Comment by rtawfik01** [comment]
@ttmtrajkovic @davorchap fyi

**Comment by rtawfik01** [comment]
Cleaned up hardcoded operands here, and removed some ifdefs: https://github.com/tenstorrent-metal/tt-metal/pull/5316

**Comment by rtawfik01** [comment]
Item 1: Removed more ifdefs here: https://github.com/tenstorrent-metal/tt-metal/pull/5375
And fixed init/config/reconfig ordering

**Comment by rtawfik01** [comment]
Item 2: Merged here #6753

**Comment by rtawfik01** [comment]
Merged item 5 here: https://github.com/tenstorrent-metal/tt-metal/pull/5491
